walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,145 @@
+    boolean[] timeSlots = new boolean[TOTAL_NUM_SLOTS];
+    Arrays.fill(timeSlots, Boolean.TRUE);
+
+    for (Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+      TimeRange when = event.getWhen();
+      int start = (when.start()) / 30;
+      int eventSlots = when.duration() / 30;
+
+      Iterator<String> eventAi = eventAttendees.iterator(); 
+      while(eventAi.hasNext()) {
+        String eventAttendee = eventAi.next();
+
+        Iterator<String> requestAi = requestAttendees.iterator(); 
+        while (requestAi.hasNext()) {
+          String requestAttendee = requestAi.next();
+          if (eventAttendee.equals(requestAttendee)) {
+            for (int i = 0; i < eventSlots; i++) {
+              timeSlots[start + i] = false;
+            }
+            break;
+          }
+        }
+      }
+    }
+    return timeSlots;
+  }
+
+  // converts boolean availability to list of available slots
+  public static ArrayList<Integer> convertToSlots (boolean timeSlots[], int numRequestSlots) {
+    ArrayList<Integer> slots = new ArrayList<Integer>();
+    boolean add;
+    for (int i = 0; i < TOTAL_NUM_SLOTS; i++) {
+      add = false;
+      if (timeSlots[i]) {
+        add = true;
+        // checks if the duration is longer than a single time slot
+        for (int j = 1; j < numRequestSlots; j++) {
+          if ((i+j) < timeSlots.length && !timeSlots[i+j]) {
+            add = false;
+          }
+        }
+      }
+      if (add) {
+        for (int j = 0; j < numRequestSlots; j++) {
+          slots.add(i+j);
+        }
+      }
+    }
+    return slots;
+  }
+
+  // converts list of available slots to a list of TimeRanges 
+  public static ArrayList<TimeRange> convertToTimes (ArrayList<Integer> slots, int numRequestSlots, int numEvents) {
+    ArrayList<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+
+    // if slots size is zero
+    if (slots.size() == 0) {
+      return possibleTimes;
+    }
+
+    int start = slots.get(0)*30;
+
+    // if there's just enough room
+    if (numRequestSlots == slots.size()) {
+      possibleTimes.add(TimeRange.fromStartEnd(start, start + numRequestSlots*30, false));
+      return possibleTimes;
+    }
+
+    // if the whole day is available
+    if (slots.size() == TOTAL_NUM_SLOTS) {
+      possibleTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+      return possibleTimes;
+    }
+
+    // everything else
+    for (int i = 0; i < slots.size() - numRequestSlots; i++) {
+      // if the next value at the next index is not +1 from the last
+      int current = slots.get(i);
+      int next = slots.get(i+1);
+      if (current != next - 1) {
+        possibleTimes.add(TimeRange.fromStartEnd(start, current*30+30, false));
+        start = next*30;
+        numEvents--;
+      }
+      if (numEvents == 0) {
+        break;
+      }
```

Comment:
okay I lost the thread we had on this but your explanation makes sense to me - maybe add a short comment since it's not immediately obvious? I think it's a good efficiency check (I suspect it's hard to tell with your test cases because they're small amounts of data to loop through either way, but your reasoning seems good to me)

Reply:
done :)